### ISSUE TITLE
How to save loss value of each sample(not the whole dataset) at every iteration?

### ISSUE THREAD

**Issue by wuzhe1234** [microsoft/LightGBM#5541]
In order to analyse the loss curve of each sample(not the whole dataset), I wrote following codes:

      loss_curve = pd.DataFrame(np.zeros((N, num_trees)))
      pred_tree = np.zeros(N, dtype=float)
      for i_tree in range(num_trees):
          pred_tree += model.predict(x_train.values, start_iteration=i_tree, num_iteration=1)
          loss_curve.iloc[:, i_tree] = self.get_loss(y_train, pred_tree)

It's time consuming. Since all the values I need has been calculated while training, is there a way to save those values while training?

**Comment by jmoralez** [comment]
Hi @wuzhe1234, thank you for your interest in LightGBM. You're right, the predictions are computed and stored internally, so  you can define a callback to store the predictions on each iteration like in the following example. Keep in mind that this isn't supported and could break in the future.
```python
import numpy as np
import lightgbm as lgb

predictions = []
def save_predictions(env):
    # have to copy because the buffer is reused in each iteration
    train_preds = env.model._Booster__inner_predict(0).copy()
    predictions.append(train_preds)


rng = np.random.RandomState(0)
X = rng.rand(100, 2)
y = rng.rand(100)
ds = lgb.Dataset(X, y)
num_iter = 5
bst = lgb.train(
    {'num_leaves': 5},
    ds,
    num_boost_round=num_iter,
    callbacks=[save_predictions],
)
for i in range(num_iter):
    computed = bst.predict(X, num_iteration=i+1)
    np.testing.assert_allclose(predictions[i], computed)
```

**Comment by wuzhe1234** [comment]
> Hi @wuzhe1234, thank you for your interest in LightGBM. You're right, the predictions are computed and stored internally, so you can define a callback to store the predictions on each iteration like in the following example. Keep in mind that this isn't supported and could break in the future.
> 
> ```python
> import numpy as np
> import lightgbm as lgb
> 
> predictions = []
> def save_predictions(env):
>     # have to copy because the buffer is reused in each iteration
>     train_preds = env.model._Booster__inner_predict(0).copy()
>     predictions.append(train_preds)
> 
> 
> rng = np.random.RandomState(0)
> X = rng.rand(100, 2)
> y = rng.rand(100)
> ds = lgb.Dataset(X, y)
> num_iter = 5
> bst = lgb.train(
>     {'num_leaves': 5},
>     ds,
>     num_boost_round=num_iter,
>     callbacks=[save_predictions],
> )
> for i in range(num_iter):
>     computed = bst.predict(X, num_iteration=i+1)
>     np.testing.assert_allclose(predictions[i], computed)
> ```

Thank you so much! Your code solve my question perfectly!
